[14.0] stock_average_daily_sale: returns handling

sebalix opened the pull request:
Not ready to be proposed on OCA, so this PR was open to ease internal reviews.

Based on:
- https://github.com/OCA/stock-logistics-reporting/pull/342

twalter-c2c committed:
stock_average_daily_sale: returns handling

Take returns into account while calculating average daily sale usage.

In our specific case, returns mean stock moves from the 'Inventory adjustment'
location to the physical stock location (of usage 'internal').

Co-Authored-By: Sébastien Alix <sebastien.alix@camptocamp.com>

twalter-c2c committed:
fixup! stock_average_daily_sale: returns handling

henrybackman commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -250,6 +315,28 @@ def _create_materialized_view(self):
+                averages_return AS(
+                    SELECT
+                        row_number() over (order by product_id) as id,
+                        concat(warehouse_id, product_id)::integer as window_id,
```
This looks dangerous to me, but perhaps context assures that there's no 11th warehouse?
```
select concat(11, 2)::integer = concat(1, 12)::integer;
-> true
```

sebalix replied:
You are right, and in the targeted customer project there are tons of warehouses, so this cast is dangerous here.

twalter-c2c replied:
Good catch! I suppose it should be fixed on the OCA as well and here only for the returns part, right @sebalix? What would be the recommended solution for that? Is something like `concat(warehouse_id, 0, product_id)` enough?

henrybackman replied:
Since record IDs are always integers, easy way to avoid conflicts is to use a non-integer character as the divider and have the combination ID be a string (unless there is a reason that this combination ID needs to be an integer?)
For example:
`concat(warehouse_id, '-', product_id)`

twalter-c2c replied:
I pushed a fix here for the returns part (new commit) and on OCA upstream https://github.com/OCA/stock-logistics-reporting/pull/357

sebalix reviewed (commented):
@twalter-c2c the new `help` should be done through a field override in customer specific modules (I guess they have been updated in that regards, not to be proposed to OCA?).

twalter-c2c committed:
stock_average_daily_sale: make sure joining on unique ids - returns

henrybackman commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -308,13 +413,19 @@ def _create_materialized_view(self):
+                            (cfg.number_days_qty_in_stock * (average_qty_by_sale - COALESCE(average_qty_by_return, 0)))
+                        ) as recommended_qty_incl_returns
                     FROM averages t
                     JOIN daily_standard_deviation ds on ds.id= t.window_id
                     JOIN stock_average_daily_sale_config cfg on cfg.id = t.config_id
                     JOIN stock_qty sqty on sqty.pp_id = t.product_id AND t.warehouse_id = sqty.warehouse_id
                     JOIN product_product pp on pp.id = t.product_id
                     JOIN product_template pt on pt.id = pp.product_tmpl_id
+                    LEFT JOIN averages_return tr ON tr.window_id = t.window_id
+                    LEFT JOIN daily_standard_deviation_return dsr on dsr.id= t.window_id
```
~~The `window_id` / `id` is not updated for all of these tables and some of the joins will not match for any record.~~

~~These have the `id` still defined as `concat(warehouse_id, product_id)::integer as window_id`~~
~~- averages~~
~~- daily_standard_deviation~~

henrybackman replied:
Ah, just now noticed the other linked OCA PR that resolves this

twalter-c2c replied:
Indeed, the solution here is only for the returns part, and that's also why tests are failing here.

twalter-c2c committed:
stock_average_daily_sale: run materialized view query for active configs only

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (59 of 59 strings)

Translation: stock-logistics-reporting-14.0/stock-logistics-reporting-14.0-stock_average_daily_sale
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-14-0/stock-logistics-reporting-14-0-stock_average_daily_sale/it/

OCA-git-bot committed:
[IMP] update dotfiles

mymage committed:
Translated using Weblate (Italian)

Currently translated at 16.6% (1 of 6 strings)

Translation: stock-logistics-reporting-14.0/stock-logistics-reporting-14.0-stock_quant_history_queued
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-14-0/stock-logistics-reporting-14-0-stock_quant_history_queued/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 25.0% (1 of 4 strings)

Translation: stock-logistics-reporting-14.0/stock-logistics-reporting-14.0-stock_account_quantity_history_location
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-14-0/stock-logistics-reporting-14-0-stock_account_quantity_history_location/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 25.0% (1 of 4 strings)

Translation: stock-logistics-reporting-14.0/stock-logistics-reporting-14.0-stock_picking_group_by_partner_by_carrier_sale_line_position
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-14-0/stock-logistics-reporting-14-0-stock_picking_group_by_partner_by_carrier_sale_line_position/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 14.2% (1 of 7 strings)

Translation: stock-logistics-reporting-14.0/stock-logistics-reporting-14.0-delivery_line_sale_line_position
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-14-0/stock-logistics-reporting-14-0-delivery_line_sale_line_position/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 16.6% (1 of 6 strings)

Translation: stock-logistics-reporting-14.0/stock-logistics-reporting-14.0-stock_picking_report_custom_description
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-14-0/stock-logistics-reporting-14-0-stock_picking_report_custom_description/it/

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -309,13 +416,19 @@ def _create_materialized_view(self):
-                        ) as recommended_qty
+                        ) as recommended_qty,
+                        GREATEST(
+                            (cfg.number_days_qty_in_stock * (average_qty_by_sale - COALESCE(average_qty_by_return, 0)) * (average_daily_sales_count - COALESCE(average_daily_returns_count, 0))) + ((ds.daily_standard_deviation - COALESCE(dsr.daily_standard_deviation, 0)) * cfg.safety_factor * sqrt(nbr_days)),
```
This new formula looks wrong to me
If you always ship 10 and you return also 10, then `average_qty_by_sale - average_qty_by_return = 0`. So a single return will put the recommended qty to 0. That's wrong.
You need to compute the daily sale - the daily return
```suggestion
                            (cfg.number_days_qty_in_stock * (
    average_qty_by_sale * average_daily_sales_count
    - COALESCE(average_qty_by_return, 0) * COALESCE(average_daily_returns_count, 0)
                            )  + ((ds.daily_standard_deviation - COALESCE(dsr.daily_standard_deviation, 0)) * cfg.safety_factor * sqrt(nbr_days)),
```

twalter-c2c replied:
@sebalix is the author of these changes, but I also know the topic so I will clarify.
We have 'hidden' the returns handling because the way we handle returns is not universal here. In this case, as returns, we consider moves where `sl_src.usage in ('inventory') AND sl_dest.usage in ('internal')` which is rather a custom requirement.
The changes you requested are still relevant in my opinion though. Thank you for the suggestion, I will fix that.

jbaudoux replied:
Thanks @twalter-c2c 
I'm still clarifying the formula with Meredith

jbaudoux reviewed (changes requested):
Why hiding such change?

You would have opened a PR on OCA, you would maybe had reviews.

twalter-c2c committed:
fixup! fixup! stock_average_daily_sale: returns handling

florentx reviewed (approved):
LG, but I did not dig into algorithm itself
